### PR TITLE
Always create full mapping SMM page table

### PR DESCRIPTION
**Pull request by td36**
# Description
In this commit, we always create full mapping SMM page table in SmmInitPageTable regardless the PCD value of PcdCpuSmmRestrictedMemoryAccess.

Previously, when PcdCpuSmmRestrictedMemoryAccess is false, only [0, 4G] is mapped in smm page table in SmmInitPageTable. If the range above 4G is accessed in SMM, SmiPFHandler will create new paging entry for the accessed range. To simplify the code logic, we also create full mapping SMM page table when PcdCpuSmmRestrictedMemoryAccess is false. Then we don't need to dynamic create paging entry for range above 4G except SMM profile is enabled.

The comparison of SMM page table before and after this change under different configuration are listed here:
1. PcdCpuSmmRestrictedMemoryAccess is TRUE
    No change except the special bits set in previous SMM paging content

2. PcdCpuSmmRestrictedMemoryAccess is FALSE and PcdCpuSmmProfileEnable is TRUE
    Before: the SMM page table when ReadyToLock covers
                 1. SMRAM range 2.SMM profile range 3. MMIO range below 4G
    After: the SMM page table when ReadyToLock covers
                 1. SMRAM range 2.SMM profile range 3. MMIO range below 4G and above 4G

3. PcdCpuSmmRestrictedMemoryAccess is FALSE and PcdCpuSmmProfileEnable is FALSE
    Before: the SMM page table when ReadyToLock covers
                 [0, 4G]
    After: the SMM page table when ReadyToLock covers
                 [0, MaxSupportPhysicalAddress]

## How This Was Tested
Tested on QEMU with OvmfPkg and QSP SimicsOpenBoardPkg with following configuration
1. PcdCpuSmmRestrictedMemoryAccess is TRUE
2. PcdCpuSmmRestrictedMemoryAccess is FALSE and PcdCpuSmmProfileEnable is TRUE
    (with PcdCpuSmmStackGuard  set to TRUE & FALSE)
3. PcdCpuSmmRestrictedMemoryAccess is FALSE and PcdCpuSmmProfileEnable is FALSE

The test results shows that:
1.All the test can boot into shell/OS.
2.The comparison of the SMM page table when ReadyToLock before and after my change are as expected.

### CODE REVIEW

**Comment by jiaxinwu** [comment]
the first patch doesn't change the special bit? 
image

File: UefiCpuPkg/PiSmmCpuDxeSmm/X64/PageTbl.c [hunk 1]
```diff
@@ -226,10 +225,8 @@ SmmInitPageTable (
 
   //
   // Generate initial SMM page table.
-  // Only map [0, 4G] when PcdCpuSmmRestrictedMemoryAccess is FALSE.
   //
-  PhysicalAddressBits = mCpuSmmRestrictedMemoryAccess ? mPhysicalAddressBits : 32;
-  PageTable           = GenSmmPageTable (mPagingMode, PhysicalAddressBits);
+  PageTable = GenSmmPageTable (mPagingMode, mPhysicalAddressBits);
```
Comment (jiaxinwu) [inline UefiCpuPkg/PiSmmCpuDxeSmm/X64/PageTbl.c:229]: Size will be the big problem if always do the full mapping with 1G paging unsupported case.
i'm not sure whether has any platform doesn't support 1g paging...

File: UefiCpuPkg/PiSmmCpuDxeSmm/X64/PageTbl.c [hunk 2]
```diff
@@ -962,13 +962,10 @@ SmiPFHandler (
     }
 
     if (mCpuSmmRestrictedMemoryAccess && IsSmmCommBufferForbiddenAddress (PFAddress)) {
-      DumpCpuContext (InterruptType, SystemContext);
```
Comment (jiaxinwu) [inline UefiCpuPkg/PiSmmCpuDxeSmm/X64/PageTbl.c:968]: This change the order of:
1. DumpCpuContext
2. DumpModuleInfoByIp

Normally, we think this is more reason we see the #PF dump code:
image

How about move the DumpModuleInfoByIp into below else logic?

File: UefiCpuPkg/PiSmmCpuDxeSmm/Ia32/SmmProfileArch.c [hunk 3]
```diff
@@ -72,3 +72,15 @@ ClearTrapFlag (
 {
   SystemContext.SystemContextIa32->Eflags &= (UINTN) ~BIT8;
 }
+
+/**
+  Create new entry in page table for page fault address in SmmProfilePFHandler.
+
+**/
+VOID
+SmiProfileMapPFAddress (
```
Comment (jiaxinwu) [inline UefiCpuPkg/PiSmmCpuDxeSmm/Ia32/SmmProfileArch.c:79]: SmiProfileXXX is not good since it's another feature not same to the SmmProfile.

how about change to SmmProfileMapPFAddress?

**Review by jiaxinwu — CHANGES_REQUESTED** [review]

File: UefiCpuPkg/PiSmmCpuDxeSmm/X64/SmmProfileArch.c [hunk 4]
```diff
@@ -353,15 +351,6 @@ RestorePageTableAbove4G (
   // If page entry does not existed in page table at all, create a new entry.
   //
   if (!Existed) {
-    if (IsAddressValid (PFAddress, &Nx)) {
```
Comment (jiaxinwu) [inline UefiCpuPkg/PiSmmCpuDxeSmm/X64/SmmProfileArch.c:210]: please also clean the IsAddressValid () function since it's not needed.

File: UefiCpuPkg/PiSmmCpuDxeSmm/SmmProfile.c [hunk 5]
```diff
@@ -1,7 +1,7 @@
 /** @file
 Enable SMM profile.
 
-Copyright (c) 2012 - 2024, Intel Corporation. All rights reserved.<BR>
+Copyright (c) 2012 - 2023, Intel Corporation. All rights reserved.<BR>
```
Comment (jiaxinwu) [inline UefiCpuPkg/PiSmmCpuDxeSmm/SmmProfile.c:4]: should be 2024

**Comment by jiaxinwu** [comment]
This is not TRUE in the first patch before you revert commit https://github.com/tianocore/edk2/commit/bef0d333dc4fccdfc75e4be31e067b467a9a4093 "UefiCpuPkg/PiSmmCpuDxeSmm:
Fix system hang when SmmProfile enable".
image

The right behavior is: 
PcdCpuSmmRestrictedMemoryAccess is FALSE and PcdCpuSmmProfileEnable is TRUE:
     Before and after : the SMM page table when ReadyToLock covers
        1. SMRAM range
        2.  SMM profile range
        3. MMIO range below 4G and above 4G

To better align with your comments, i suggest revert commit https://github.com/tianocore/edk2/commit/bef0d333dc4fccdfc75e4be31e067b467a9a4093 "UefiCpuPkg/PiSmmCpuDxeSmm:
Fix system hang when SmmProfile enable" first.

File: UefiCpuPkg/PiSmmCpuDxeSmm/SmmProfile.c [hunk 6]
```diff
@@ -1345,18 +1320,17 @@ SmmProfilePFHandler (
     //
     // If SMM profile does not start, call original page fault handler.
     //
-    SmiDefaultPFHandler ();
+    SmiProfileMapPFAddress ();
```
Comment (jiaxinwu) [inline UefiCpuPkg/PiSmmCpuDxeSmm/SmmProfile.c:1323]: please consider whether need call this function if SMM profile does not start since we create the full page table.

File: UefiCpuPkg/PiSmmCpuDxeSmm/X64/PageTbl.c [hunk 7]
```diff
@@ -226,31 +225,29 @@ SmmInitPageTable (
...
-  PageTable           = GenSmmPageTable (mPagingMode, PhysicalAddressBits);
+  PageTable = GenSmmPageTable (mPagingMode, mPhysicalAddressBits);
+
+  if (FeaturePcdGet (PcdCpuSmmProfileEnable)) {
+    if (m5LevelPagingNeeded) {
+      Pml5Entry = (UINT64 *)PageTable;
+      //
+      // Set Pml5Entry sub-entries number for smm PF handler usage.
+      //
+      SetSubEntriesNum (Pml5Entry, 1);
```
[the remark this answers is not in the document]
Comment (td36) [reply UefiCpuPkg/PiSmmCpuDxeSmm/X64/PageTbl.c:237]: It's for dynamic creating page table.

"do we understand why original code sets the Pml5Entry SubEntries to 2?"
I think it's not needed when mCpuSmmRestrictedMemoryAccess is TRUE.

For now, it's only needed when SMM profile is enabled

File: UefiCpuPkg/PiSmmCpuDxeSmm/Ia32/PageTbl.c [hunk 8]
```diff
@@ -196,7 +196,8 @@ SmiPFHandler (
       );
   } else {
     DumpCpuContext (InterruptType, SystemContext);
-    SmiDefaultPFHandler ();
+    CpuDeadLoop ();
+    goto Exit;
```
[the remark this answers is not in the document]
Comment (td36) [reply UefiCpuPkg/PiSmmCpuDxeSmm/Ia32/PageTbl.c:185]: When DEBUG, we can use debugger to jump out the CpuDeadLoop

File: UefiCpuPkg/PiSmmCpuDxeSmm/X64/PageTbl.c (hunk 1 above)
Comment (td36) [reply UefiCpuPkg/PiSmmCpuDxeSmm/X64/PageTbl.c:229]: I tested the PR in some Intel SIMICS platform and it worked. Could you specify any client platform that doesn't support 1G page table size?

File: UefiCpuPkg/PiSmmCpuDxeSmm/X64/SmmProfileArch.c (hunk 4 above)
Comment (td36) [reply UefiCpuPkg/PiSmmCpuDxeSmm/X64/SmmProfileArch.c:210]: it's still used in RestorePageTableAbove4G() logic, please check. I didn't change the logic of that part

File: UefiCpuPkg/PiSmmCpuDxeSmm/Ia32/PageTbl.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (td36) [reply UefiCpuPkg/PiSmmCpuDxeSmm/Ia32/PageTbl.c:185]: Should be removed. Since the next line of code is Exit

File: UefiCpuPkg/PiSmmCpuDxeSmm/SmmProfile.c (hunk 5 above)
Comment (td36) [reply UefiCpuPkg/PiSmmCpuDxeSmm/SmmProfile.c]: Updated all copyright year

File: UefiCpuPkg/PiSmmCpuDxeSmm/Ia32/SmmProfileArch.c (hunk 3 above)
Comment (td36) [reply UefiCpuPkg/PiSmmCpuDxeSmm/Ia32/SmmProfileArch.c]: Changed to SmmProfileMapPFAddress

File: UefiCpuPkg/PiSmmCpuDxeSmm/X64/PageTbl.c (hunk 2 above)
Comment (td36) [reply UefiCpuPkg/PiSmmCpuDxeSmm/X64/PageTbl.c:968]: moved the DumpModuleInfoByIp into below else logic

File: UefiCpuPkg/PiSmmCpuDxeSmm/SmmProfile.c (hunk 6 above)
Comment (td36) [reply UefiCpuPkg/PiSmmCpuDxeSmm/SmmProfile.c]: Added a new commit to remove it

**Comment by td36** [comment]
> the first patch doesn't change the special bit? if yes, please correct the commit message:
> 
> image

Updated the commit message

**Comment by td36** [comment]
> This is not TRUE in the first patch before you revert commit [bef0d33](https://github.com/tianocore/edk2/commit/bef0d333dc4fccdfc75e4be31e067b467a9a4093) "UefiCpuPkg/PiSmmCpuDxeSmm: Fix system hang when SmmProfile enable". image
> 
> The right behavior before revert the patch is: **Before and after : the SMM page table when ReadyToLock covers 1. SMRAM range 2. SMM profile range 3. MMIO range below 4G and above 4G**
> 
> To better align with your comments, i suggest revert commit [bef0d33](https://github.com/tianocore/edk2/commit/bef0d333dc4fccdfc75e4be31e067b467a9a4093) "UefiCpuPkg/PiSmmCpuDxeSmm: Fix system hang when SmmProfile enable" first.

Moved it to the first commit

File: UefiCpuPkg/PiSmmCpuDxeSmm/X64/PageTbl.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (jiaxinwu) [reply UefiCpuPkg/PiSmmCpuDxeSmm/X64/PageTbl.c:237]: could you explain why need 2 sub-entries of Pml5Entry for dynamic creating page table?
SetSubEntriesNum (Pml5Entry, 1) means set the SubEntries to 2.

File: UefiCpuPkg/PiSmmCpuDxeSmm/X64/PageTbl.c (hunk 1 above)
Comment (jiaxinwu) [reply UefiCpuPkg/PiSmmCpuDxeSmm/X64/PageTbl.c:229]: i don't know. i just raise the concern here to make sure we have considered the case. 

if size is not enough, i checked the code, it should hang at the failure of AllocatePageTableMemory in GenSmmPageTable.

File: UefiCpuPkg/PiSmmCpuDxeSmm/Ia32/PageTbl.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (jiaxinwu) [reply UefiCpuPkg/PiSmmCpuDxeSmm/Ia32/PageTbl.c]: ok.

File: UefiCpuPkg/PiSmmCpuDxeSmm/SmmProfile.c (hunk 5 above)
Comment (jiaxinwu) [reply UefiCpuPkg/PiSmmCpuDxeSmm/SmmProfile.c]: ok

File: UefiCpuPkg/PiSmmCpuDxeSmm/SmmProfile.c (hunk 6 above)
Comment (jiaxinwu) [reply UefiCpuPkg/PiSmmCpuDxeSmm/SmmProfile.c]: ok

File: UefiCpuPkg/PiSmmCpuDxeSmm/Ia32/SmmProfileArch.c (hunk 3 above)
Comment (jiaxinwu) [reply UefiCpuPkg/PiSmmCpuDxeSmm/Ia32/SmmProfileArch.c]: ok

File: UefiCpuPkg/PiSmmCpuDxeSmm/X64/SmmProfileArch.c (hunk 4 above)
Comment (jiaxinwu) [reply UefiCpuPkg/PiSmmCpuDxeSmm/X64/SmmProfileArch.c:210]: i checked the commit message:

image

we think it must be invalid address run into the RestorePageTableAbove4G(), then below code is unnecessary. what's the reason we keep that?

image

File: UefiCpuPkg/PiSmmCpuDxeSmm/X64/PageTbl.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (td36) [reply UefiCpuPkg/PiSmmCpuDxeSmm/X64/PageTbl.c:237]: It's related with the page reclaim mechanism. If the subentry num of a non-leaf PDPDE is 0, when the child of this non-leaf entry is reclaimed, the non-leaf parent may also be relaimed since the only present child is reclaimed.

At first, I thought that it's set to non-zero value to prevent the PML4 paging structure pointered by PLM5[0] from being reclaimed in the page reclaim mechanism. This is to make sure [0, 4G] is always present in SMM page table

But I checked code and found that there is no logic to reclaim any PML4 buffer pointered by a PML5 entry. So I think this line of code to set subentry for PML5 is not needed. But it's not related with the task and doesn't impact anything. I think we can move it in our future plan that we may totally replace the page reclaim mechanism with a new and simpler mechanism

File: UefiCpuPkg/PiSmmCpuDxeSmm/X64/SmmProfileArch.c (hunk 4 above)
Comment (td36) [reply UefiCpuPkg/PiSmmCpuDxeSmm/X64/SmmProfileArch.c:210]: The page table that covers 4G above range may be reclaimed in ReclaimPages(). Let's keep the  IsAddressValid () and IsValidPFAddress in RestorePageTableAbove4G()

Comment (jiaxinwu) [reply UefiCpuPkg/PiSmmCpuDxeSmm/X64/SmmProfileArch.c]: agree, the MMIO above 4G range can be reclaimed, still need IsAddressValid () check.

**Comment by jiaxinwu** [comment]
agree we need IsAddressValid(), since the MMIO above 4G range is not marked with IA32_PG_PMNT. So, it can be reclaimed that might lead to the original Present to Non-present in page table. So, still need IsAddressValid () check. could you add new commit in this PR to descript the purpose of  IsAddressValid ()? Becasue the  IsAddressValid usage case or cover scope has been changed after this PR.

Besides, within IsAddressValid (), below code can be cleaned also, i suggest to add one more PR to remove below code:
image


To summarize my comment, add additional 2 commits:
1. clarify the IsAddressValid usage case or cover scope
2. clean some unused code in the IsAddressValid 

what do you think?

**Comment by jiaxinwu** [comment]
refine below commit message, it doesn't related to Pml5Entry:

image

**Comment by td36** [comment]
> refine below commit message, it doesn't related to Pml5Entry:
> 
> image

Updated the commit message. Pls check

**Comment by td36** [comment]
> agree we need IsAddressValid(), since the MMIO above 4G range is not marked with IA32_PG_PMNT. So, it can be reclaimed that might lead to the original Present to Non-present in page table. So, still need IsAddressValid () check. could you add new commit in this PR to descript the purpose of IsAddressValid ()? Becasue the IsAddressValid usage case or cover scope has been changed after this PR.
> 
> Besides, within IsAddressValid (), below code can be cleaned also, i suggest to add one more PR to remove below code: image
> 
> To summarize my comment, add additional 2 commits:
> 
> 1. clarify the IsAddressValid usage case or cover scope
> 2. clean some unused code in the IsAddressValid
> 
> what do you think?

Added a new commit to rename and simplify it. Pls check

**Review by jiaxinwu — APPROVED** [review]

**Review by jiaxinwu — APPROVED** [review]

**Review by niruiyu — APPROVED** [review]
